contractcourt/channel_arbitrator.go

```diff
@@ -1269,15 +1340,254 @@ func (c *ChannelArbitrator) isPreimageAvailable(hash lntypes.Hash) (bool,
+		height, trigger, confHTLCs,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	// With this actions computed, we'll now check the diff of the HTLCs on
+	// the commitments, and cancel back any that are on the pending but not
+	// the non-pending.
+	remoteDiffActions := c.checkRemoteDiffActions(
+		height, activeHTLCs, pendingConf,
+	)
+
+	// Finally, we'll merge all the chain actions and the final set of
+	// chain actions.
+	remoteCommitActions.Merge(remoteDiffActions)
+	return remoteCommitActions, nil
+}
+
+// checkRemoteDiffActions checks the set difference of the HTLCs on the remote
+// confirmed commit and remote dangling commit for HTLCS that we need to cancel
+// back. If we find any HTLCs on the remote pending but not the remote, then
+// we'll mark them to be failed immediately.
+func (c *ChannelArbitrator) checkRemoteDiffActions(height uint32,
+	activeHTLCs map[HtlcSetKey]htlcSet,
+	pendingConf bool) ChainActionMap {
+
+	// First, we'll partition the HTLCs into those that are present on the
+	// confirmed commitment, and those on the dangling commitment.
+	confHTLCs := activeHTLCs[RemoteHtlcSet]
+	danglingHTLCs := activeHTLCs[RemotePendingHtlcSet]
+	if pendingConf {
+		confHTLCs = activeHTLCs[RemotePendingHtlcSet]
+		danglingHTLCs = activeHTLCs[RemoteHtlcSet]
+	}
+
+	// Next, we'll create a set of all the HTLCs confirmed commitment.
+	remoteHtlcs := make(map[uint64]struct{})
+	for _, htlc := range confHTLCs.outgoingHTLCs {
+		remoteHtlcs[htlc.HtlcIndex] = struct{}{}
+	}
+
+	// With the remote HTLCs assembled, we'll mark any HTLCs only on the
+	// remote dangling commitment to be failed asap.
+	actionMap := make(ChainActionMap)
+	for _, htlc := range danglingHTLCs.outgoingHTLCs {
+		if _, ok := remoteHtlcs[htlc.HtlcIndex]; ok {
+			continue
+		}
+
+		actionMap[HtlcFailNowAction] = append(
+			actionMap[HtlcFailNowAction], htlc,
+		)
+
+		log.Tracef("ChannelArbitrator(%v): immediately failing "+
+			"htlc=%x from remote commitment",
+			c.cfg.ChanPoint, htlc.RHash[:])
+	}
+
+	return actionMap
+}
+
+// constructChainActions returns the set of actions that should be taken for
+// confirmed HTLCs at the specified height. Our actions will depend on the set
+// of HTLCs that were active across all channels at the time of channel
+// closure.
+func (c *ChannelArbitrator) constructChainActions(confCommitSet *CommitSet,
+	height uint32, trigger transitionTrigger) (ChainActionMap, error) {
+
+	// If we've reached this point and have not confirmed commitment set,
+	// then this is an older node that had a pending close channel before
+	// the CommitSet was introduced. In this case, we'll just return the
+	// existing ChainActionMap they had on disk.
+	if confCommitSet == nil {
+		return c.log.FetchChainActions()
+	}
+
+	// Otherwise we have the full commitment set written to disk, and can
+	// proceed as normal.
+	htlcSets := confCommitSet.toActiveHTLCSets()
+	switch *confCommitSet.ConfCommitKey {
```

Comment:
Make it is an idea to take out the call to `checkCommitChainActions` from `checkLocalChainActions` and `checkRemoteChainActions` and put it here. It is clear then that we always check the confirmed commit and use the switch only to merge the 'extra' actions.

Reply:
It's just a matter of following the function to see that `checkCommitChainActions` is called first. It also states something along these lines in the `godoc` comments of each of the methods.